File: ai/genetic.py
import numpy as np
from .neural_network import NeuralNetwork

class GeneticAlgorithm:
    def __init__(self, population_size=10, top_units=4):
        self.population_size = population_size
        self.top_units = top_units
        self.generation = 1
        self.mutation_rate = 0.2
        
        self.best_fitness = 0
        self.best_network = None
        self.best_generation = 0
        
        self.population = []
        self.create_population()
    
    def create_population(self):
        """Create initial population of neural networks"""
        self.population = []
        for _ in range(self.population_size):
            self.population.append(NeuralNetwork())
    
    def evolve_population(self, birds):
        """Evolve population based on fitness scores"""
        sorted_birds = sorted(birds, key=lambda x: x.fitness, reverse=True)
        
        best_bird = sorted_birds[0]
        if best_bird.fitness > self.best_fitness:
            self.best_fitness = best_bird.fitness
            self.best_generation = self.generation
            self.best_network = self.population[best_bird.index].copy()
            print(f"New best fitness: {self.best_fitness:.0f} in generation {self.generation}")
        
        winners = [self.population[bird.index] for bird in sorted_birds[:self.top_units]]
        
        new_population = []
        
        new_population.extend(winners)
        
        while len(new_population) < self.population_size:
            if len(new_population) == self.top_units:
                parent1, parent2 = winners[0], winners[1]
            else:
                parent1 = np.random.choice(winners)
                parent2 = np.random.choice(winners)
            
            offspring = self.crossover(parent1, parent2)
            offspring = self.mutate(offspring)
            new_population.append(offspring)
        
        self.population = new_population
        self.generation += 1
        
        if self.mutation_rate == 1.0 and sorted_birds[0].fitness > 0:
            self.mutation_rate = 0.2
    
    def crossover(self, parent1, parent2):
        """Perform crossover between two parent networks"""
        child = NeuralNetwork()
        weights1 = parent1.get_weights()
        weights2 = parent2.get_weights()
        
        new_weights = {}
        for key in weights1:
            shape = weights1[key].shape
            mask = np.random.rand(*shape) > 0.5
            new_weights[key] = np.where(mask, weights1[key], weights2[key])
        
        child.set_weights(new_weights)
        return child
    
    def mutate(self, network):
        """Apply random mutations to the network"""
        weights = network.get_weights()
        
        for key in weights:
            mutation_mask = np.random.rand(*weights[key].shape) < self.mutation_rate
            mutation = np.random.randn(*weights[key].shape) * 0.1
            weights[key] = np.where(mutation_mask, weights[key] + mutation, weights[key])
        
        network.set_weights(weights)
        return network
    
    def activate_brain(self, bird, pipe):
        """Get neural network output for a bird"""
        if not pipe:
            return
        
        horizontal_distance = (pipe.x + pipe.width - bird.x) / 300
        height_difference = (bird.y - (pipe.gap_y + pipe.gap_height/2)) / 200
        velocity = bird.velocity / 8
        
        network = self.population[bird.index]
        output = network.forward([horizontal_distance, height_difference, velocity])
           
        bird.current_inputs = [horizontal_distance, height_difference, velocity]
        bird.current_output = output
        
        if output > 0.5:
            bird.flap() 

File: static/js/genetic.js
class GeneticAlgorithm {
    constructor(populationSize = 50, mutationRate = 0.2) {
        this.populationSize = populationSize;
        this.mutationRate = mutationRate;
        this.generation = 1;
        this.bestFitness = 0;
        this.population = [];
        
        // Initialize mutation rates
        this.initialMutationRate = 0.25;
        this.normalMutationRate = 0.1;
        this.eliteSize = 2;
        
        this.pendingPromises = new Map();
        
        // Create initial population
        this.createPopulation();
        
        // Initialize worker after population is created
        this.initWorker();
    }
    
    initWorker() {
        if (this.worker) {
            this.worker.terminate();
        }
        
        try {
            this.worker = new Worker('/static/js/ai.worker.js');
            
            this.worker.onmessage = (e) => {
                const { type, data } = e.data;
                const resolve = this.pendingPromises.get(type);
                if (resolve) {
                    resolve(data);
                    this.pendingPromises.delete(type);
                }
            };
            
            this.worker.onerror = (error) => {
                console.error('Worker error:', error);
                // Reinitialize worker on error
                setTimeout(() => this.initWorker(), 1000);
            };
        } catch (error) {
            console.error('Failed to initialize worker:', error);
        }
    }
    
    createPopulation() {
        console.log('Creating new population...');
        this.population = [];
        for (let i = 0; i < this.populationSize; i++) {
            const network = new NeuralNetwork();
            // Verify network initialization
            if (!network.weights1 || !network.weights2) {
                console.error('Network not properly initialized');
            }
            this.population.push(network);
        }
        console.log(`Created population of ${this.population.length} networks`);
    }
    
    crossover(parent1, parent2) {
        const child = new NeuralNetwork();
        
        // Improved crossover with weighted averaging
        const crossoverValue = (v1, v2) => {
            const r = Math.random();
            return v1 * r + v2 * (1 - r);  // Weighted average
        };
        
        // Crossover weights and biases
        child.weights1 = parent1.weights1.map((row, i) => 
            row.map((w, j) => crossoverValue(w, parent2.weights1[i][j]))
        );
        
        child.weights2 = parent1.weights2.map((row, i) => 
            row.map((w, j) => crossoverValue(w, parent2.weights2[i][j]))
        );
        
        child.bias1 = parent1.bias1.map((b, i) => 
            crossoverValue(b, parent2.bias1[i])
        );
        
        child.bias2 = parent2.bias2.map((b, i) => 
            crossoverValue(b, parent2.bias2[i])
        );
        
        return child;
    }
    
    mutate(network) {
        const mutationStrength = 0.5;  // Increased from 0.2
        
        // Mutate weights with improved probability distribution
        const mutateValue = (value) => {
            if (Math.random() < this.mutationRate) {
                // Use Gaussian distribution for mutations
                const gaussian = () => {
                    let u = 0, v = 0;
                    while (u === 0) u = Math.random();
                    while (v === 0) v = Math.random();
                    return Math.sqrt(-2.0 * Math.log(u)) * Math.cos(2.0 * Math.PI * v);
                };
                return value + gaussian() * mutationStrength;
            }
            return value;
        };

        // Apply mutations with higher probability in early generations
        const earlyGenBonus = Math.max(0, 5 - this.generation) * 0.1;
        this.mutationRate = this.normalMutationRate + earlyGenBonus;

        // Apply mutations
        network.weights1 = network.weights1.map(row => row.map(mutateValue));
        network.weights2 = network.weights2.map(row => row.map(mutateValue));
        network.bias1 = network.bias1.map(mutateValue);
        network.bias2 = network.bias2.map(mutateValue);
        
        return network;
    }
    
    async activateBrain(bird, pipe) {
        if (!pipe) return;
        
        const horizontalDistance = (pipe.x + pipe.width - bird.x) / 300;
        const heightDifference = (bird.y - (pipe.gapY + pipe.gapHeight/2)) / 200;
        const velocity = bird.velocity / 8;
        
        bird.currentInputs = [horizontalDistance, heightDifference, velocity];
        
        const network = this.population[bird.index];
        if (!network) return;
        
        // Use Web Worker for forward pass
        try {
            const result = await this.workerForward(bird.currentInputs, network);
            bird.currentOutput = result.output;
            network.lastHiddenActivations = result.lastHiddenActivations;
            
            if (result.output > 0.5) {
                bird.flap();
            }
        } catch (error) {
            console.error('Error in activateBrain:', error);
        }
    }
    
    workerForward(inputs, network) {
        return new Promise((resolve, reject) => {
            if (!this.worker) {
                console.log('Initializing new worker...');
                this.initWorker();
            }
            
            try {
                const messageData = {
                    type: 'FORWARD',
                    data: {
                        inputs: inputs,
                        weights1: network.weights1,
                        weights2: network.weights2,
                        bias1: network.bias1,
                        bias2: network.bias2
                    }
                };
                
                this.pendingPromises.set('FORWARD_RESULT', resolve);
                this.worker.postMessage(messageData);
            } catch (error) {
                console.error('Error in workerForward:', error);
                reject(error);
            }
        });
    }
    
    async evolvePopulation(sortedBirdIndexes) {
        // Validate inputs
        if (!Array.isArray(sortedBirdIndexes) || sortedBirdIndexes.length === 0) {
            console.error('Invalid bird indexes for evolution');
            return this.population;
        }

        // Adjust mutation rate based on generation
        this.mutationRate = this.generation <= 3 ? this.initialMutationRate : this.normalMutationRate;
        
        // Create a clean copy of the population data for the worker
        const populationData = this.population.map(network => ({
            weights1: network.weights1.map(row => [...row]),
            weights2: network.weights2.map(row => [...row]),
            bias1: [...network.bias1],
            bias2: [...network.bias2]
        }));

        return new Promise((resolve, reject) => {
            if (!this.worker) {
                this.initWorker();
            }
            
            try {
                this.pendingPromises.set('EVOLVE_RESULT', (newPopulation) => {
                    // Convert the plain objects back to NeuralNetwork instances
                    const evolvedPopulation = newPopulation.map(networkData => {
                        const network = new NeuralNetwork();
                        network.weights1 = networkData.weights1;
                        network.weights2 = networkData.weights2;
                        network.bias1 = networkData.bias1;
                        network.bias2 = networkData.bias2;
                        return network;
                    });
                    resolve(evolvedPopulation);
                });

                this.worker.postMessage({
                    type: 'EVOLVE',
                    data: {
                        population: populationData,
                        sortedBirdIndexes,
                        populationSize: this.populationSize,
                        mutationRate: this.mutationRate
                    }
                });
            } catch (error) {
                console.error('Error in evolvePopulation:', error);
                reject(error);
            }
        });
    }
    
    cleanup() {
        if (this.worker) {
            this.worker.terminate();
            this.worker = null;
        }
        this.pendingPromises.clear();
    }
} 